pkg/adaptor/controller.go

```diff
@@ -0,0 +1,365 @@
+		return nil
+	}
+
+	serviceFQDN := c.toFQDN(ts.Spec.Service, ts.Namespace)
+	// Check if the Service Profile is already present
+	sp, err := c.spclientset.LinkerdV1alpha2().ServiceProfiles(ts.Namespace).Get(ctx, serviceFQDN, metav1.GetOptions{})
+	if err != nil {
+		// Return if its not a not found error
+		if !errors.IsNotFound(err) {
+			log.Errorf("couldn't retrieve serviceprofile/%s: %s", serviceFQDN, err)
+			return err
+		}
+
+		// Create a Service Profile resource as it does not exist
+		sp, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(ts.Namespace).Create(ctx, c.toServiceProfile(ts), metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+		log.Infof("created serviceprofile/%s for trafficsplit/%s", sp.Name, ts.Name)
+	} else {
+		log.Infof("serviceprofile/%s already present", sp.Name)
+		if ignoreAnnotationPresent(sp) {
+			log.Infof("skipping update of serviceprofile/%s as ignore annotation is present", sp.Name)
+			return nil
+		}
+
+		// Check if SP Matches the TS, and update if it not
+		spFromTs := c.toServiceProfile(ts)
+		if !equal(spFromTs, sp) {
+			log.Infof("serviceprofile/%s does not match trafficscplit/%s", sp.Name, ts.Name)
+			updateDstOverrides(sp, ts, c.clusterDomain)
+			_, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(ts.Namespace).Update(ctx, sp, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+			log.Infof("updated serviceprofile/%s as it's not equivalent to trafficsplit/%s", sp.Name, ts.Name)
+		}
+	}
+
+	return nil
+}
+
+func ignoreAnnotationPresent(sp *serviceprofile.ServiceProfile) bool {
+	_, ok := sp.Annotations[ignoreServiceProfileAnnotation]
+	return ok
+}
+
+func equal(spA *serviceprofile.ServiceProfile, spB *serviceprofile.ServiceProfile) bool {
+	if spA.Name != spB.Name {
+		return false
+	}
+
+	if spA.Namespace != spB.Namespace {
+		return false
+	}
+
+	if len(spA.Spec.DstOverrides) != len(spB.Spec.DstOverrides) {
+		return false
+	}
+
+	dstOverridesA := make(map[string]string)
+	for _, dstA := range spA.Spec.DstOverrides {
+		dstOverridesA[dstA.Authority] = dstA.Weight.String()
+	}
+
+	// Check if all the authorties from spB exist
+	// in dstOverridesA with the same weight
+	for _, dstB := range spB.Spec.DstOverrides {
+		weight, ok := dstOverridesA[dstB.Authority]
+		if !ok {
+			return false
+		}
+
+		if weight != dstB.Weight.String() {
+			return false
+		}
+	}
+
+	return true
+}
+
+// enqueueTS takes a Ts resource and converts it into a key
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than TS.
+func (c *SMIController) enqueueTS(obj interface{}) {
+	var key string
+	var err error
+	if key, err = trafficSplitKeyFunc(obj); err != nil {
+		log.Error(err)
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+// trafficSplitKeyFunc takes a TS, and gives back a
+// namespace/name/service key string
+func trafficSplitKeyFunc(obj interface{}) (string, error) {
```

Comment:
Instead of using a string for the key, we should use a (namespace, name, service) struct.  
Or, even better, can we just use the TS itself as the key?

Reply:
We can do that but I don't see the benefit here, The purpose of the key being just a string(we need namespace, name and service to match with the relevant serviceprofile from the trafficsplit) is that, we can move the event around without having to worry about the object. and to just use the TS itself as the key, when the value is also the TS? 🤔